### ISSUE TITLE
Bug: [prefer-const] False positives when using increment operator with non-null assertion operator

### ISSUE THREAD

**Issue by janek515** [typescript-eslint/typescript-eslint#10270]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.3&fileType=.ts&code=DYUwLgBAhhC8EEZoGcIDsCuBbARiAThAD4QZoAmIAZgJZojkDcAsAFBQCEA1F40A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQA7UQDNFoBaAYwHsA7ZAF3ShOkuknDAF8ROg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=true

### Repro Code

```TypeScript
let a = 1 as number | undefined;
a!++;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "prefer-const": "error",
  },
};
```

### Expected Result

It should not error.

### Actual Result

ESLint shows a message `'a' is never reassigned. Use 'const' instead. 1:5 - 1:6`

### Additional Info

Similar to #2804
